### PR TITLE
feat(presets): group `@types/react` and `@types/react-dom` together, along with other React packages

### LINKED ISSUE #15081: Bundle React Types with React Monorepo
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.14.1

### Please select which platform you are using if self-hosting.

GitHub Enterprise Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

3.2.11

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Use the `config:base` preset, a lot of React dependencies are bundled:
- React
- React Dom
- React Test Render
- React Is
- etc.

But their types are not all bundled with them:
- [@types/react-test-renderer](https://github.com/DefinitelyTyped/DefinitelyTyped)
- [@types/react](https://github.com/DefinitelyTyped/DefinitelyTyped)
- [@types/react-dom](https://github.com/DefinitelyTyped/DefinitelyTyped)

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository, I can create one if you like.

My Renovate Config:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":prHourlyLimitNone"
  ]
}
```

### PR DESCRIPTION
**Pull request by astorije**
## Changes

This new group bundles upgrades of `@types/react` and `@types/react-dom` together, as discussed [here](https://github.com/renovatebot/renovate/issues/15081). It avoids situations like these:

Screen Shot 2023-02-16 at 2 49 19 PM

## Context

- Closes #15081

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by HonkingGoose** [comment]
Hi @astorije, thank you for your PR. ❤️ 

Please stop force pushing to your pull request branch. It makes reviewing your work harder for the maintainers.

Please also read [Our contributing guide, do not force push to your pull request branch](https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#do-not-force-push-to-your-pull-request-branch) to learn why we want people to avoid force-pushing.

I'll let the maintainers review your work. 😉

**Comment by astorije** [comment]
Ah yeah absolutely, I'm well aware, just my PR was definitely not ready to review / not being reviewed so I let myself rebase to fix it. Now that it's under review, no more force-pushing :)
Thank you!

File: lib/config/presets/internal/group.ts
Comment (rarkins) [inline lib/config/presets/internal/group.ts:343]: I don't think this will work, if it is based on `sourceUrl`, because it will AND together matchSourceUrls and matchPackageNames

**Review by rarkins — CHANGES_REQUESTED** [review]
Is the goal here to create a new group rule which users must opt into, or to combine them into the existing "React monorepo" group which comes with config:base?

**Comment by astorije** [comment]
Hi @rarkins, thanks for reaching out so quickly as usual :)

> Is the goal here to create a new group rule which users must opt into, or to combine them into the existing "React monorepo" group which comes with config:base?

The latter, the default, base preset should bundle `@types/react` and `@types/react-dom` together in the same PR (as opposed to my screenshot above), and if I understand the linked issue correctly, the consensus was to bundle them with the React upgrades themselves.

Then I sort of followed the steps [here](https://github.com/renovatebot/renovate/issues/15081#issuecomment-1152885583) to create a new group, so as to not alter the existing "React monorepo" preset since, technically speaking, the `@types/` dependencies aren't part of the monorepo.

That said, I contribute to Renovate so infrequently that I am not 100% familiar with the syntax so apologies if I got it wrong (and it's a shame, I love this tool so much, I wish I had more time to contribute!).

File: lib/config/presets/internal/group.ts
Comment (astorije) [reply lib/config/presets/internal/group.ts:343]: Oh interesting, I was under the imprsesion it would do an OR.

What would be the way to do that then? Essentially group together packages from the `"https://github.com/facebook/react"` `sourceURL` and packages matching `@types/react` and `@types/react-dom`?

Also we don't use it ourselves so I forgot about it, but the linked issue also refers to `@types/react-test-renderer`. I should probably include it to the PR for completeness once we reach the correct syntax (and as a reminder, we can't use `@types/react` as a prefix because there are a lot of them, mostly not matching packages from the React monorepo).

**Comment by rarkins** [comment]
In this case the `groupName` needs to match with `group:reactMonorepo`'s

File: lib/config/presets/internal/group.ts
Comment (astorije) [inline lib/config/presets/internal/group.ts:344]: @rarkins do you mean this?

```suggestion
        groupName: 'react monorepo',
```

I got this from https://docs.renovatebot.com/presets-group/#groupreactmonorepo

Comment (rarkins) [reply lib/config/presets/internal/group.ts:344]: Yes, I think so

Comment (rarkins) [reply lib/config/presets/internal/group.ts:344]: And then this can be added to `group:recommended` maybe

Comment (astorije-vmware) [reply lib/config/presets/internal/group.ts:344]: I think I did, could you confirm/advise?

Comment (astorije) [reply lib/config/presets/internal/group.ts:344]: I think I did, could you confirm/advise?

File: lib/config/presets/internal/group.ts
Comment (rarkins) [inline lib/config/presets/internal/group.ts:343]:
```suggestion
```

File: lib/config/presets/internal/group.ts
Comment (rarkins) [reply lib/config/presets/internal/group.ts:344]: Yes

**Review by rarkins — CHANGES_REQUESTED** [review]

**Review by rarkins — APPROVED** [review]

**Comment by HonkingGoose** [comment]
@rarkins if you want to merge this PR, you'll need to resolve one last review comment. 😉

**Pull request merged**
